Upgrade to Discord.js v13

Hellysonrp opened the pull request:
This PR upgrades from `discord.js` v12 to v13 (#351).

Opening as a draft because I couldn't get `@discordjs/voice` to work.
It simply doesn't detect the installed libraries (`@discordjs/opus` and `libsodium-wrappers`).

## The problem
I don't have much experience with node, but here is what I found about the problem:
node loads `@discordjs/voice` from the `index.mjs` file, inside `node_modules`. This file has a variable `O` that represents the `require` function/object. There is a function in this file that gets the version of another library and returns it; this function uses `O.resolve` to get the path to the module (`require.resolve`). This throws an exception (`O.resolve is not a function`). Here is the code I'm talking about:
![image](https://user-images.githubusercontent.com/5438762/139977858-320993d1-98e0-43f1-9903-b5d6c7a9dd31.png)
This exception makes it that `not found` is always returned.
Even if I hack the code to use `path.resolve` (from `node:path`, represented by the `ve` function in the minified code) to get the path, there is another exception thrown after that, because the code tries to `require` the `package.json` from the library, resulting in an `Uncaught Error: Dynamic require of "<path>" is not supported`.
I genuinely tried everything I could to get it to recognize the required libraries, but it simply doesn't. I don't know if this has something to do with the way this project is compiled (i.e. the tsconfig options, which I tried to change too, with no avail) or if it is a bug in some other thing (node, the library, or my environment), but it is annoying 😩  
I've tested with Windows and WSL (Debian). Same results.
It isn't only this function that doesn't detect the libraries, because if I ignore this and try to listen to something, it throws an exception because it can't find the "encryption library" (`libsodium-wrappers`).

I couldn't test the changes because of that.

## The changes

### Packages
I've upgraded some libraries in an attempt to get it to work. I don't know if all the upgraded libraries are required to make it work.
In any case, here is the required upgrades (AFAIK):
* Minimum nodejs engine version bumped to `16.0.0` (minimum version supported by `@discordjs/voice`)
* `discord.js` version bumped to v13
* Installed `@discord/voice`
* Installed `libsodium-wrappers` (`@discordjs/voice` required dependency; I've tried to install `sodium`, but couldn't get it to install on Windows)

Any other change in the `package.json` are my attempts to get it to work, and maybe aren't required.

### Code
* Added the Intents to the client (required in `discord.js` v13)
* Changed some events
* Changed some `discord.js` function calls
* Renamed `dispatcher` to `audioPlayer` in `player.ts` to better represent the underlying type
* Added `debug(generateDependencyReport());` to the client `ready` event, to see if `@discordjs/voice` succeeded to load the dependencies

Some changes that might be worth to do are:
* Use the `AudioPlayer` to directly load the streams and cached files, since it is compatible with FFmpeg (it's listed as an 'optional dependency')
    * If I understood it correctly, this would make the `fs-capacitor` dependency obsolete, which is good, because it has a [memory leak](https://prnt.sc/1xrhsou).
* Better `queue` command (like the one Groovy had), with buttons to see all pages of the queue
* Slash commands (`discord.js` v13 added a way to register slash commands)

## Final considerations
I've put a lot of hours trying to get it to work.
I have no idea how to get it to work.
Any help is appreciated.

And just to properly link the issue: fixes #351

codetheweb commented:
Thank you for putting so much effort into this (and the accompanying detailed report)!

I spent some time debugging and it appears that `@discordjs/voice` isn't currently playing nice with ES modules (what we're outputting and using) so I [opened a new issue](https://github.com/discordjs/voice/issues/227). We'll have to wait for that to be fixed before we can move forward.

---------	

> Use the AudioPlayer to directly load the streams and cached files, since it is compatible with FFmpeg (it's listed as an 'optional dependency') 

What do you mean by this? We're already using AudioPlayer (`createAudioPlayer()`)

> Better queue command (like the one Groovy had), with buttons to see all pages of the queue

I haven't used Groovy before but if by that you mean adding buttons to the embed to move back / forward that makes sense. Feel free to open a new issue.

> Slash commands (discord.js v13 added a way to register slash commands)

See https://github.com/codetheweb/muse/issues/331

Hellysonrp commented:
> What do you mean by this? We're already using AudioPlayer (`createAudioPlayer()`)

What I meant is that we could use the `createAudioResource` function to read the cached files and the YouTube stream directly, but now I see that I completely missed the fact that `fs-capacitor` is used to write the cached files too...
My mind was in another place when I was writing the draft :man_facepalming:
Sorry about that :sweat_smile:

> I haven't used Groovy before but if by that you mean adding buttons to the embed to move back / forward that makes sense. Feel free to open a new issue.

Yes, exactly that. I'll open it after this PR is (hopefully) merged.
I might even develop the command, but because of IRL, I'm not having much time to work in this bot (that's one reason I didn't finish my other draft).

---

I'll wait until the issue in `@discordjs/voice` is solved, then I'll test my changes and finish this draft.
Thank you for your time and this amazing bot.

Hellysonrp commented:
I've added [the workaround mentioned in the issue](https://github.com/discordjs/voice/issues/227#issuecomment-964707120).
It seems to run without problems.
The only thing that bothers me is that the workaround imports the functions as `any`, and eslint doesn't like that.

Let me know if any changes are needed.

codetheweb commented:
It looks like the issue was partly fixed on Discord.js's side, although they're still working on it: https://github.com/discordjs/voice/pull/232.

Let's wait until they fully fix it and then I'll review this. Thanks again for your work.

codetheweb commented:
It looks like this should be resolved now in [v0.7.4](https://github.com/discordjs/voice/releases/tag/v0.7.4) if you wanna try it out. :)

Hellysonrp commented:
I've updated to the latest version of `@discordjs/voice` and updated the `Lint & build` action to use nodejs 16, since this version is the minimum required now.

I've briefly tested the bot, and everything seems to be working.
My "production" bot has been updated to use this code. Since it's used daily for about 10 hours a day, any issues in the basic functionality, if any, will probably be caught until next weekend.
I have noted some issues not related to this change and will be opening them separately.

codetheweb commented:
Great!

It seems to mostly be working for me. However, I tried setting it up again from scratch (kicked it and re-added it to my guild) and had some issues. Selecting a channel using emojis works fine but it seems to stall at the second step asking for a prefix. 

If it's working for you I can debug on my end, just super busy currently with school.

Hellysonrp commented:
I've found the issue. An intent was missing.
I've added it and it works now.

codetheweb commented:
All seems to be working now. :)

Thank you!
